clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordBatch.java

```diff
@@ -206,11 +209,21 @@ public double compressionRate() {
     }
 
     public boolean isFull() {
-        return recordsBuilder.isFull();
+        return !isWritable || recordsBuilder.isFull();
+    }
+
+    public void closeWithTransactionState(long pid, short epoch, int baseSequence) {
+        recordsBuilder.closeWithTransactionState(pid, epoch, baseSequence);
+        isWritable = false;
```

Comment:
Not sure why we need this additional variable. Could we just modify the logic of `isWritable` below to check that the builder is not close AND still have space then, which seems simpler?